ipfscluster_test.go

```diff
@@ -816,6 +816,275 @@ func TestClustersReplication(t *testing.T) {
+	if nClusters < 5 {
+		t.Skip("Need at least 5 peers")
+	}
+
+	clusters, mock := createClusters(t)
+	defer shutdownClusters(t, clusters, mock)
+	for _, c := range clusters {
+		c.config.ReplicationFactorMin = 1
+		c.config.ReplicationFactorMax = nClusters
+	}
+
+	h, _ := cid.Decode(test.TestCid1)
+	err := clusters[0].Pin(api.PinCid(h))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Shutdown two peers
+	clusters[nClusters-1].Shutdown()
+	clusters[nClusters-2].Shutdown()
+
+	time.Sleep(time.Second) // let metric expire
+
+	waitForLeader(t, clusters)
+
+	// allow metrics to arrive to new leader
+	delay()
+
+	err = clusters[0].Pin(api.PinCid(h))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	p, err := clusters[0].PinGet(h)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if len(p.Allocations) != nClusters {
+		t.Error("allocations should still be nCluster even if not all available")
+	}
+
+	if p.ReplicationFactorMax != nClusters {
+		t.Error("rplMax should have not changed")
+	}
+}
+
+// This test checks tat repinning something that has becomed
+// underpinned does re-allocations when it's not sufficiently
+// pinned anymore
+func TestClustersReplicationMinMaxRealloc(t *testing.T) {
+	if nClusters < 5 {
+		t.Skip("Need at least 5 peers")
+	}
+
+	clusters, mock := createClusters(t)
+	defer shutdownClusters(t, clusters, mock)
+	for _, c := range clusters {
+		c.config.ReplicationFactorMin = 3
+		c.config.ReplicationFactorMax = 4
+	}
+
+	h, _ := cid.Decode(test.TestCid1)
+	err := clusters[0].Pin(api.PinCid(h))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	delay()
+
+	p, err := clusters[0].PinGet(h)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	firstAllocations := p.Allocations
+
+	peerIDMap := make(map[peer.ID]*Cluster)
+	for _, a := range clusters {
+		peerIDMap[a.id] = a
+	}
+
+	// kill two of the allocations
+	alloc1 := peerIDMap[firstAllocations[0]]
+	alloc2 := peerIDMap[firstAllocations[1]]
+	safePeer := peerIDMap[firstAllocations[2]]
+
+	alloc1.Shutdown()
+	alloc2.Shutdown()
+
+	time.Sleep(time.Second) // let metric expire
+
+	waitForLeader(t, clusters)
+
+	// allow metrics to arrive to new leader
+	delay()
+
+	// Repin - (although this might have been taken of if there was an alert
+	err = safePeer.Pin(api.PinCid(h))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	p, err = safePeer.PinGet(h)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	secondAllocations := p.Allocations
+
+	if len(secondAllocations) != minInt(nClusters-2, 4) {
+		t.Error("pin should be allocated again to a few peers")
+	}
+
+	strings1 := api.PeersToStrings(firstAllocations)
+	strings2 := api.PeersToStrings(secondAllocations)
+	sort.Strings(strings1)
+	sort.Strings(strings2)
+	t.Logf("Allocs1: %s", strings1)
+	t.Logf("Allocs2: %s", strings2)
+
+	if fmt.Sprintf("%s", strings1) == fmt.Sprintf("%s", strings2) {
+		t.Error("allocations should have changed")
+	}
+
+	if len(secondAllocations) < 3 {
+		t.Error("allocations should be more than rplMin")
+	}
+}
+
```

Comment:
The last two tests hit the `needed <= 0` and `nCandidatesValid < needed` branches of obtainAllocations.  It doesn't look like we have a test for `wanted <=0` yet, i.e. repin with a smaller replication factor which would complete the set.

Reply:
Hmm yeah ok